### ISSUE TITLE
not doing autoregressive decoding for sequence sampling?

### ISSUE THREAD

**Issue by onenoc** [oxpig/AntiFold#14]
I believe that there is a mistake in how sampling is done: it is not using autoregressive decoding conditional on the sequence sampled so far.

In line 297 of antifold/antiscripts.py, you compute the logits in a single forward pass. Later, you use the logits from a single forward pass for sampling. However, this is not doing standard autoregressive decoding, as it conditions the sampled token at step k on the original sequence up to k-1, while autoregressive decoding would condition the sampled token at step k on the sampled sequence up to k-1. What you're doing is fine for training, but not for sampling. If you look at antifold/esm/inverse_folding/gvp_transformer.py, in the sample function (line 93), we see that starting at line 133 they decode one token at a time, doing L forward passes. This conditions the next sampled token on the sampled sequence so far.

In the paper, you have sequence recovery for the CDRH3 region of the original sequence. ESM-IF1 would then be computed using the context of the sequence sampled so far, while antifold would be computed using the context of the original sequence. This would give antifold a large advantage in terms of potentially more useful context.

**Comment by mhoie** [comment]
Hi onenoc, yes we have provided a function for unconditional sampling (single forward pass) as an example sampling function here and on the webserver. We do accept pull requests for added functionality :) However as you mention, one can easily re-use the ESM-IF1 sampling function to auto-regressively sample.

For the sequence recovery plots, probabilities for all models are extracted with a single forward pass (unconditionally) and not auto-regressively. The models are thus on equal terms here.

**Comment by onenoc** [comment]
Thanks for the clarification, this is helpful. Is there a domain specific reason (e.g. something about antibodies) for doing it this way vs the classic autoregressive sampling? The esm-IF and as best as I can tell the ProteinMPNN repos do it the classical way.
